src/types/infer.ts

```diff
@@ -2092,4 +2007,413 @@ export class InferType {
 
         return new ErrorType(def.name, argTs);
     }
+
+    /**
+     * Computes the function type for the public accessor corresponding to a state variable
+     */
+    getterFunType(v: VariableDeclaration): FunctionType {
+        const [args, ret] = this.getterArgsAndReturn(v);
+
+        return new FunctionType(
+            v.name,
+            args,
+            ret instanceof TupleType ? ret.elements : [ret],
+            FunctionVisibility.External,
+            FunctionStateMutability.View
+        );
+    }
+
+    /**
+     * Computes the argument types and return type for the public accessor
+     * corresponding to a state variable.
+     */
+    getterArgsAndReturn(v: VariableDeclaration): [TypeNode[], TypeNode] {
+        const argTypes: TypeNode[] = [];
+
+        let type = v.vType;
+
+        assert(
+            type !== undefined,
+            "Called getterArgsAndReturn() on variable declaration without type",
+            v
+        );
+
+        while (true) {
+            if (type instanceof ArrayTypeName) {
+                argTypes.push(new IntType(256, false));
+
+                type = type.vBaseType;
+            } else if (type instanceof Mapping) {
+                argTypes.push(
+                    this.typeNameToSpecializedTypeNode(type.vKeyType, DataLocation.Memory)
+                );
+
+                type = type.vValueType;
+            } else {
+                break;
+            }
+        }
+
+        let retType = this.typeNameToSpecializedTypeNode(type, DataLocation.Memory);
+
+        if (
+            retType instanceof PointerType &&
+            retType.to instanceof UserDefinedType &&
+            retType.to.definition instanceof StructDefinition
+        ) {
+            const elements: TypeNode[] = [];
+
+            for (const member of retType.to.definition.vMembers) {
+                const memberT = member.vType;
+
+                assert(
+                    memberT !== undefined,
+                    "Unexpected untyped struct member",
+                    retType.to.definition
+                );
+
+                if (memberT instanceof Mapping || memberT instanceof ArrayTypeName) {
```

Comment:
I wonder what should we do if `memberT` is a struct?

Reply:
We don't have to fix this right now, cause it is not making the behavior worse. Just wondering what the correct semantic is. Will make a note of this in an issue and add it here.

Reply:
Tracked under https://github.com/ConsenSys/solc-typed-ast/issues/169